fuzzy-native/src/MatcherBase.cpp

```diff
@@ -35,12 +35,51 @@ inline string str_to_lower(const std::string &s) {
   return lower;
 }
 
+float distance_to_root(const std::string *root, const std::string *value) {
+  size_t num_common_dirs = 0;
+  size_t i = 0;
+
+  // Count number of common directories
+  for (; i < root->length() && i < value->length(); i++) {
+    if (root->at(i) != value->at(i)) {
+      break;
+    }
+    if (root->at(i) == '/' || root->at(i) == '\\') {
+      num_common_dirs++;
+    }
+  }
+
+  if (i == root->length()) {
+    return 0;
+  }
+
+  // Count total number of directories in root
+  size_t num_dirs_root = num_common_dirs;
+  for (size_t j = i; j < root->length(); j++) {
+    if (root->at(j) == '/' || root->at(j) == '\\') {
+      num_dirs_root++;
+    }
+  }
+
+  float distance = num_dirs_root - num_common_dirs;
+  float penalty = 1;
+  for (size_t j = i; j < value->length(); j++) {
+    if (value->at(j) == '/' || value->at(j) == '\\') {
+      distance += penalty;
+      penalty *= 2.0;
+    }
+  }
+
+  return distance;
```

Comment:
So, this is just my gut feeling, but I think that if the root is `/a/b/c/d` then

1. naturally `/a/b/c/xxx` > `/a/b/xxx` > `/a/xxx` (as you've already captured)
2. but also `/a/b/c/x/y/z` should still be better than `/a/b/x`.

I think in case 2) the "extra path" penalty unfairly penalizes `/a/b/c/x/y/z` and makes `/a/b/x` better - but I feel like this usually wouldn't be the case. Sharing more of the prefix feels strictly better.

Facebook example: if you're searching for a BUCK file you might be in `fbcode/my_product/test.cpp`.
I think `fbcode/my_product/sub_product/test/BUCK` should still be better than `fbcode/x/BUCK`, but I think here the exponential penalty would end up selecting `fbcode/x/BUCK`.

I think that prioritizing `num_common_dirs` and only then ordering by the number of slashes in `value` (total dirs) is better; it's actually better as a tuple `(num_common_dirs, num_total_dirs)`. You could alternatively use a very high multiplier on `num_common_dirs`; e.g. `100 * num_common_dirs + num_total_dirs`.